guide/adopting-the-configuration-cache.md

```diff
@@ -0,0 +1,384 @@
+            execSpec.workingDir(project.getProjectDir());
+            execSpec.setCommandLine("git", "describe", "--tags", "--abbrev=0");
+            execSpec.setStandardOutput(output);
+         });
+         return output.toString().strip();
+      } catch (Exception e) {
+         throw new RuntimeException(e);
+      }
+   }
+}
+```
+
+### Fixing "invocation of `Task.project` at execution time is unsupported"
+
+#### Before
+- `MyTask` is already a [Gradle managed type](managed-types-and-properties.md)
+- `MyTask` is calling `getProject()` at build time, causing the configuration cache to fail: **invocation of `Task.project` at execution time is unsupported**
+
+```java
+abstract class MyTask extends DefaultTask {
+   @InputFile
+   protected abstract FileProperty getMyFiles();
+
+   @TaskAction
+   public void delete() {
+      boolean deleted = getProject().delete(getMyFiles().getFiles());
+      // Do something with `deleted`...
+      Path projectDir = getProject().getProjectDir().toPath();
+      // Do something with `projectDir`...
+   }
+}
+
+```
+
+#### After
+- To get the project dir, we can use the `ProjectLayout` service.
+- To do file system operations like delete, we can use the `FileSystemOperations` service
+- We can inject these services into `MyTask`
+
+```java
+abstract class MyTask extends DefaultTask {
+   @InputFile
+   protected abstract FileProperty getMyFiles();
+
+   @Inject
+   protected abstract ProjectLayout getProjectLayout();
+
+   @Inject
+   protected abstract FileSystemOperations getFileSystemOperations();
+
+   @TaskAction
+   public void delete() {
+      WorkResult result = getFileSystemOperations().delete(spec -> spec.delete(getMyFiles().getFiles()));
+      // Do something with `result`...
+      Path projectDir = getProjectLayout().getProjectDirectory().getAsFile().toPath();
+      // Do something with `projectDir`...
+   }
+}
+```
+
+
+## Two strategies to write Configuration Cache friendly Gradle
+
+To summarize the two strategies we used to solve Configuration Cache issues:
+
+### You can inject services into Gradle-managed types
+
+Do you need the project directory? Inject [`ProjectLayout`](https://docs.gradle.org/current/javadoc/org/gradle/api/file/ProjectLayout.html). Do you need to run a bash command? Inject [`ExecOperations`](https://docs.gradle.org/current/javadoc/org/gradle/process/ExecOperations.html). Gradle offers a [list](https://docs.gradle.org/current/userguide/service_injection.html) of injectable services which will cover 99% of your use cases. However, that list is incomplete — in reality,  almost [anything in Gradle source](https://github.com/search?q=repo%3Agradle%2Fgradle%20%40ServiceScope&type=code) annotated with [`@ServiceScope`](https://github.com/gradle/gradle/blob/196bb409d47f5b6e39d62edd39be939f7606a5cc/platforms/core-runtime/stdlib-java-extensions/src/main/java/org/gradle/internal/service/scopes/ServiceScope.java#L43) can be injected into a Gradle managed type.
```

Comment:
* as stated above, the bash command is context dependent, are you in a task action? are you in a plugin code?